meshmode/mesh/__init__.py

```diff
@@ -1199,59 +1270,178 @@ def boundary_tag_bit(boundary_tag):
-                neighbors=face_id_pairs[1].elements[is_neighbor_adj],
-                neighbor_faces=face_id_pairs[1].faces[is_neighbor_adj])
+            grp_list.append(
+                InteriorAdjacencyGroup(
+                    igroup=igrp,
+                    ineighbor_group=i_neighbor_grp,
+                    elements=face_id_pairs[0].elements[is_neighbor_adj],
+                    element_faces=face_id_pairs[0].faces[is_neighbor_adj],
+                    neighbors=face_id_pairs[1].elements[is_neighbor_adj],
+                    neighbor_faces=face_id_pairs[1].faces[is_neighbor_adj]))
             face_has_neighbor[
                 face_id_pairs[0].faces[is_neighbor_adj],
                 face_id_pairs[0].elements[is_neighbor_adj]] = True
 
         has_bdry = not np.all(face_has_neighbor)
         if has_bdry:
-            faces, elements = np.where(~face_has_neighbor)
-            element_faces = faces.astype(face_id_dtype)
-            neighbors = np.full(len(elements),
-                -(boundary_tag_bit(BTAG_ALL)
-                    | boundary_tag_bit(BTAG_REALLY_ALL)),
-                dtype=element_id_dtype)
+            bdry_indices = np.where(~face_has_neighbor)
+            bdry_element_faces = bdry_indices[0].astype(face_id_dtype)
+            bdry_elements = bdry_indices[1].astype(element_id_dtype)
+            belongs_to_bdry = np.full(
+                (len(boundary_tags), len(bdry_elements)), False)
+
             if face_vertex_indices_to_tags is not None:
-                for i in range(len(elements)):
-                    ref_fvi = grp.face_vertex_indices()[element_faces[i]]
-                    fvi = frozenset(grp.vertex_indices[elements[i], ref_fvi])
+                for i in range(len(bdry_elements)):
+                    ref_fvi = grp.face_vertex_indices()[bdry_element_faces[i]]
+                    fvi = frozenset(grp.vertex_indices[bdry_elements[i], ref_fvi])
                     tags = face_vertex_indices_to_tags.get(fvi, None)
                     if tags is not None:
-                        tag_mask = 0
                         for tag in tags:
-                            tag_mask |= boundary_tag_bit(tag)
-                        neighbors[i] = -((-neighbors[i]) | tag_mask)
-            neighbor_faces = np.zeros(len(elements), dtype=face_id_dtype)
-            grp_map[None] = FacialAdjacencyGroup(
-                igroup=igrp,
-                ineighbor_group=None,
-                elements=elements,
-                element_faces=element_faces,
-                neighbors=neighbors,
-                neighbor_faces=neighbor_faces)
-
-        facial_adjacency_groups.append(grp_map)
+                            btag_idx = boundary_tag_to_index[tag]
+                            belongs_to_bdry[btag_idx, i] = True
+
+            for btag_idx, btag in enumerate(boundary_tags):
+                indices, = np.where(belongs_to_bdry[btag_idx, :])
+                if len(indices) > 0:
```

Comment:
This should be sufficient too:
```suggestion
                if indices:
```

Reply:
I don't think I can do these. When I try, I get:

```python
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
Oh apologies! Then disregard this and related comments if they don't actually work!